### PR TITLE
Support for shared code variables

### PR DESCRIPTION
**Pull request by webrouse**
Part of: https://keboola.atlassian.net/browse/COM-1138

Prvy commit, implementacia, ma `300` riadkov, zvysok su testy.

### CODE REVIEW

File: internal/pkg/mapper/sharedcode/load.go [hunk 1]
```diff
@@ -13,7 +13,47 @@ type loader struct {
 	errors    *utils.Error
 }
 
-// AfterLocalLoad - load shared code from filesystem to target config.
+// MapAfterRemoteLoad - extract shared code "variables_id".
+func (m *sharedCodeMapper) MapAfterRemoteLoad(recipe *model.RemoteLoadRecipe) error {
```
Comment (webrouse) [inline internal/pkg/mapper/sharedcode/load.go:17]: Nacitanie `variables_id` pri `remote load`.

File: internal/pkg/mapper/sharedcode/persist.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
+package sharedcode
+
+import (
+	"fmt"
+
+	"github.com/keboola/keboola-as-code/internal/pkg/model"
+)
+
+func (m *sharedCodeMapper) MapBeforePersist(recipe *model.PersistRecipe) error {
```
Comment (webrouse) [inline internal/pkg/mapper/sharedcode/persist.go:9]: Vytvorenie `SharedCodeVariablesForRelation` pri persist.

File: internal/pkg/mapper/sharedcode/save.go [hunk 3]
```diff
@@ -16,6 +16,42 @@ type writer struct {
 	errors    *utils.Error
 }
 
+// MapBeforeRemoteSave - add "variables_id" to shared code.
+func (m *sharedCodeMapper) MapBeforeRemoteSave(recipe *model.RemoteSaveRecipe) error {
```
Comment (webrouse) [inline internal/pkg/mapper/sharedcode/save.go:20]: Ulozenie `variables_id` pred `remote save`.

File: internal/pkg/model/naming.go [hunk 4]
```diff
@@ -185,6 +185,10 @@ func (n Naming) ConfigPath(parentPath string, component *Component, config *Conf
 	case parent.IsConfig() && component.IsScheduler():
 		template = string(n.SchedulerConfig)
 	case parent.IsConfig() && component.IsVariables():
+		// Regular component with variables
+		template = string(n.VariablesConfig)
+	case parent.IsConfigRow() && component.IsVariables() && parentKey.(ConfigRowKey).ComponentId == SharedCodeComponentId:
```
Comment (webrouse) [inline internal/pkg/model/naming.go:190]: Uprava namingu.

File: internal/pkg/model/naming.go [hunk 5]
```diff
@@ -354,6 +359,13 @@ func (n Naming) MatchConfigPath(parent Kind, path PathInProject) (componentId st
 		}
 	}
 
+	// Shared code variables, parent is config row
+	if parent.IsConfigRow() && parentKey.(ConfigRowKey).ComponentId == SharedCodeComponentId {
```
Comment (webrouse) [inline internal/pkg/model/naming.go:363]: Uprava namingu - `MatchConfigPath` pri operacii `persist` zistuje, ci sa jedna o config.

File: internal/pkg/model/sharedcode.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
+package model
+
+import (
+	"fmt"
+)
+
+// SharedCodeVariablesForRelation - variables for shared code.
+type SharedCodeVariablesForRelation struct {
```
Comment (webrouse) [inline internal/pkg/model/sharedcode.go:8]: `SharedCodeVariablesForRelation`/`SharedCodeVariablesFromRelation`

Su podobne s `VariablesForRelation/VariablesFromRelation`:
- `SharedCodeVariablesFromRelation` je definovana na `config row`
- a typ komponenty musi byt shared code.

File: internal/pkg/plan/persist.go [hunk 7]
```diff
@@ -93,6 +94,10 @@ func (b *persistPlanBuilder) tryAdd(fullPath string, parent model.RecordPaths) b
 		if b.tryAddConfig(path, parent.ConfigKey) != nil {
 			return true
 		}
+	case *model.ConfigRowState:
+		if b.tryAddConfig(path, parent.ConfigRowKey) != nil {
```
Comment (webrouse) [inline internal/pkg/plan/persist.go:98]: config row (shared code) moze mat vo svojom adresari config (variables).

File: internal/pkg/plan/persist.go [hunk 8]
```diff
@@ -106,13 +111,13 @@ func (b *persistPlanBuilder) tryAdd(fullPath string, parent model.RecordPaths) b
...
 			}
-			if action := b.tryAddConfig(path, parent.Key); action != nil {
-				// Set ConfigId on config persist, now it is unknown
-				parent.OnPersist = append(parent.OnPersist, func(parentKey model.Key) {
-					action.ParentKey = parentKey
-				})
-				return true
-			}
+		}
+		if action := b.tryAddConfig(path, parent.Key); action != nil {
```
Comment (webrouse) [inline internal/pkg/plan/persist.go:115]: Config row moze obsahovat config, preto ide tento blok z podmienky von.

File: internal/pkg/mapper/sharedcode/load_test.go [hunk 9]
```diff
@@ -8,8 +8,44 @@ import (
 	"github.com/keboola/keboola-as-code/internal/pkg/filesystem"
 	. "github.com/keboola/keboola-as-code/internal/pkg/mapper/sharedcode"
 	"github.com/keboola/keboola-as-code/internal/pkg/model"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils"
 )
 
+func TestSharedCodeMapAfterRemoteLoad(t *testing.T) {
```
Comment (webrouse) [inline internal/pkg/mapper/sharedcode/load_test.go:14]: Unit test - remote load.

File: internal/pkg/mapper/sharedcode/persist_test.go [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	. "github.com/keboola/keboola-as-code/internal/pkg/mapper/sharedcode"
+	"github.com/keboola/keboola-as-code/internal/pkg/model"
+)
+
+func TestSharedCodeMapBeforePersist(t *testing.T) {
```
Comment (webrouse) [inline internal/pkg/mapper/sharedcode/persist_test.go:12]: Unit test - persist.

File: internal/pkg/mapper/sharedcode/save_test.go [hunk 11]
```diff
@@ -8,8 +8,51 @@ import (
 	"github.com/keboola/keboola-as-code/internal/pkg/filesystem"
 	. "github.com/keboola/keboola-as-code/internal/pkg/mapper/sharedcode"
 	"github.com/keboola/keboola-as-code/internal/pkg/model"
+	"github.com/keboola/keboola-as-code/internal/pkg/utils"
 )
 
+func TestSharedCodeMapBeforeRemoteSave(t *testing.T) {
```
Comment (webrouse) [inline internal/pkg/mapper/sharedcode/save_test.go:14]: Unit test -remote save.

File: internal/pkg/plan/persist_test.go [hunk 12]
```diff
@@ -425,6 +425,223 @@ func TestPersistSharedCode(t *testing.T) {
 	tc.run(t)
 }
 
+func TestPersistSharedCodeWithVariables(t *testing.T) {
```
Comment (webrouse) [inline internal/pkg/plan/persist_test.go:428]: Test `persist` operacie.

File: internal/pkg/plan/persist_test.go [hunk 13]
```diff
@@ -425,6 +425,223 @@ func TestPersistSharedCode(t *testing.T) {
...
+			"main/_shared",
+			"main/_shared/keboola.python-transformation-v2",
+			"main/_shared/keboola.python-transformation-v2/codes",
+			"main/_shared/keboola.python-transformation-v2/codes/my-code",
+			"main/_shared/keboola.python-transformation-v2/codes/my-code/code.py",
+			"main/_shared/keboola.python-transformation-v2/codes/my-code/config.json",
+			"main/_shared/keboola.python-transformation-v2/codes/my-code/description.md",
+			"main/_shared/keboola.python-transformation-v2/codes/my-code/meta.json",
+			"main/_shared/keboola.python-transformation-v2/codes/my-code/variables",
+			"main/_shared/keboola.python-transformation-v2/codes/my-code/variables/config.json",
```
Comment (webrouse) [inline internal/pkg/plan/persist_test.go:454]: Shared code s variables.

File: test/pull/expected-stdout [hunk 14]
```diff
@@ -4,14 +4,14 @@ INFO	Plan for "pull" operation:
 INFO	  +  B bar
 INFO	  +  C bar/_shared/keboola.python-transformation-v2
 INFO	  +  R bar/_shared/keboola.python-transformation-v2/codes/code-with-variables
+INFO	  +  C bar/_shared/keboola.python-transformation-v2/codes/code-with-variables/variables
```
Comment (webrouse) [inline test/pull/expected-stdout:7]: Upraveny `pull` test, variables su v adresari pod shared code.

File: test/pull/out/.keboola/manifest.json [hunk 15]
```diff
@@ -54,6 +54,20 @@
         }
       ]
     },
+    {
+      "branchId": %%TEST_BRANCH_BAR_ID%%,
+      "componentId": "keboola.variables",
+      "id": "%%TEST_BRANCH_BAR_CONFIG_SHARED_CODE_VARIABLES_ID%%",
+      "path": "variables",
+      "relations": [
```
Comment (webrouse) [inline test/pull/out/.keboola/manifest.json:62]: Presunute vyssie, zmena `path` + `sharedCodeVariablesFor` relation.

File: test/push-shared-code-with-vars/expected-stdout [hunk 16]
```diff
@@ -0,0 +1,7 @@
+Plan for "push" operation:
+  +  C my-branch/_shared/keboola.python-transformation-v2
+  +  R my-branch/_shared/keboola.python-transformation-v2/codes/code-with-variables
+  +  C my-branch/_shared/keboola.python-transformation-v2/codes/code-with-variables/variables
```
Comment (webrouse) [inline test/push-shared-code-with-vars/expected-stdout:4]: Novy test pre `push`.

File: test/push-shared-code-with-vars/in/.keboola/manifest.json [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+    {
+      "id": %%TEST_BRANCH_MAIN_ID%%,
+      "path": "my-branch"
+    }
+  ],
+  "configurations": [
+    {
+      "branchId": %%TEST_BRANCH_MAIN_ID%%,
+      "componentId": "keboola.shared-code",
+      "id": "%%TEST_NEW_TICKET_1%%",
```
Comment (webrouse) [inline test/push-shared-code-with-vars/in/.keboola/manifest.json:31]: Ak chcem v teste nove ID, tak mozem pouzit `TEST_NEW_TICKET_x` ... pri prvom vyskyte sa vygeneruje nove ID pomocou API, .... pri dalsich vyskytoch sa iba nahradi (to tu uz bolo, len vysvetlenie).

File: test/push-shared-code-with-vars/expected-state.json [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+          },
+          "rows": [
+            {
+              "name": "Code with variables",
+              "description": "test fixture",
+              "changeDescription": "Updated from #KeboolaCLI",
+              "isDisabled": false,
+              "configuration": {
+                "code_content": "num1 = {{num1}}\nnum2 = {{num2}}\nsum = num1 + num2\n",
+                "variables_id": "%%TEST_NEW_TICKET_5%%"
```
Comment (webrouse) [inline test/push-shared-code-with-vars/expected-state.json:45]: Push test, `variables_id` sa pridalo a ulozilo.

**Review by webrouse — COMMENTED** [review]
- shared code je `config row`, napr. https://github.com/keboola/keboola-as-code/tree/main/test/pull/out/bar/_shared/keboola.python-transformation-v2/codes/code-with-variables
- shared code moze mat `variables_id` ... tj. premenne, ktore pouzia
- ale nema (default) `variables values`

Na prvy pohlad je to podobne s `variables`, ale:
- tie su definovane pre `config` nie pre `config row`
- preto shard codes variables musia reprezentovat nove `relations`,  ... aj ked v API je to ulozene pod `keboola.variables` komponentou.

**Review by JakubMatejka — APPROVED** [review]
👍

**Pull request merged**